quickwit/quickwit-common/src/rate_limited_tracing.rs

```diff
@@ -17,44 +17,121 @@
+            call_count: (val & ((1 << 32) - 1)) as u32,
         }
+    }
+}
+
+impl From<Count> for u64 {
+    fn from(count: Count) -> u64 {
+        ((count.generation as u64) << 32) + count.call_count as u64
+    }
+}
+
+/// Helper function used in [`rate_limited_tracing`] to determine if this line should log,
+/// and update the related counters.
+pub fn should_log<F: Fn() -> Instant>(
+    count_atomic: &AtomicU64,
+    last_reset_atomic: &AtomicU64,
+    limit: u32,
+    now: F,
+) -> bool {
+    //  count_atomic is treated as 2 u32: upper bits count "generation", lower bits count number of
+    //  calls since LAST_RESET. We assume there won't be 2**32 calls to this log in ~60s.
+    //  Generation is free to wrap arround.
+
+    let count = count_atomic.fetch_add(1, Ordering::Acquire);
+    if count == 0 {
+        // this can only be reached the very 1st time we log
+        last_reset_atomic.store(now().as_ticks(), Ordering::Release);
+    }
 
-        let do_log = if count >= $limit {
-            let mut last_reset = LAST_RESET.lock().unwrap();
-            let current_time = Instant::now();
-            let should_reset = last_reset
-                .map(|last_reset| current_time.duration_since(last_reset) >= Duration::from_secs(60))
-                .unwrap_or(true);
-
-            if should_reset {
-                *last_reset = Some(current_time);
-                // we store 1 because we are already about to log something
-                COUNT.store(1, Ordering::Relaxed);
-                true
+    let count: Count = count.into();
+
+    if count.call_count < limit {
+        return true;
+    }
+
+    let current_time = Duration::from_ticks(now().as_ticks());
+    let last_reset = Duration::from_ticks(last_reset_atomic.load(Ordering::Acquire));
+
+    let should_reset = current_time.abs_diff(last_reset) >= Duration::from_secs(60);
+
+    if !should_reset {
+        // we are over-limit and not far enough in time to reset: don't log
+        return false;
+    }
+
+    let mut update_time = false;
+    let mut can_log = false;
+
+    let _ = count_atomic.fetch_update(Ordering::Release, Ordering::Acquire, |current_count| {
+        let mut current_count: Count = current_count.into();
+        if count.generation == current_count.generation {
+            // we can update generation&time, so we can definitely log
+            update_time = true;
+            can_log = true;
+            let new_count = Count {
+                generation: count.generation.wrapping_add(1),
+                call_count: 1,
+            };
+            Some(new_count.into())
+        } else {
+            // we can't update generation&time, but maybe we can still log?
+            update_time = false;
+            if current_count.call_count < limit {
+                // we can log, update the count
+                can_log = true;
+                current_count.call_count += 1;
+                Some(current_count.into())
```

Comment:
i can't create a proper suggestion due to a UI glitch


nitpick. I think you derive can_log from the result.
```
    let update_res = count_atomic.fetch_update(Ordering::Release, Ordering::Acquire, |current_count| {
        let mut current_count: Count = current_count.into();
        if count.generation == current_count.generation {
            // we can update generation&time, so we can definitely log
            let new_count = Count {
                generation: count.generation.wrapping_add(1),
                call_count: 1,
            };
            Some(new_count.into())
        } else {
            // Generation seems to have already been updated. Maybe we can still log?
            update_time = false;
            if current_count.call_count < limit {
                // we can log, update the count
                current_count.call_count += 1;
                Some(current_count.into())
}
..
      let can_log: bool = update_res.is_ok();
```

Reply:
hum, that looks right. In a previous revision I didn't do the "save some contention by not recording that we tried to log" so we would always get Ok(_), but now that it's done, an Err happens iff we can't log